doc/guide/dts: DTS documentation add-on (#13429)
zephyrproject-rtos/zephyr: doc/guides/dts/index.rst

architectures.

The device tree is a data structure for dynamically describing hardware
using a Device Tree Source (DTS) data structure language. It allows to avoid
hard coding every detail of a board's hardware into the operating system.

In Linux, it is compiled into a compact Device Tree Blob (DTB) using a Device
Tree Compiler (DTC), then the hardware-describing DTB is passed to the operating
system at boot time. This allows the same compiled Linux kernel to support
different hardware configurations within an architecture family (e.g., ARM,
x86, PowerPC) and moves a significant part of the hardware description out of
the kernel binary itself.

> Storing the DTB and the code required to parse it requires an amount of memory

Author: Actually, the first proposed sentence is somewhat unclear to the non native speaker I am. What about : 
```
For larger systems, the benefit of the offered flexibility overcomes the requested footprint overhead. [..]
```

> Storing the DTB and the code required to parse it requires an amount of memory

Reviewer: i like what @dbkinder better ;)

> Storing the DTB and the code required to parse it requires an amount of memory

Author: > i like what @dbkinder better ;)

Ok, let's go with that then!

architectures.

The device tree is a data structure for dynamically describing hardware
using a Device Tree Source (DTS) data structure language. In Linux, it is
compiled into a compact Device Tree Blob (DTB) using a Device Tree Compiler
(DTC). Rather than hard coding every detail of a board's hardware into the
operating system, the hardware-describing DTB is passed to the operating
system at boot time. This allows the same compiled Linux kernel to support
different hardware configurations within an architecture family (e.g., ARM,
x86, PowerPC) and moves a significant part of the hardware description out of
the kernel binary itself.

Traditional usage of device tree involves storing of the Device Tree Blob.
The DTB is then used during runtime for configuration of device drivers.
However, storing the DTB and the code required to parse it requires an amount
of footprint that cannot be afforded on micro-controllers systems with limited
resources that are the primary target of Zephyr.
Hence, in Zephyr, the DTS information will be used only during compile time.
Information about the system is extracted from the compiled DTS and used to

> create the application image.

Reviewer 2: When I read this it sounds like two paragraphs of information that then abruptly ends with; "but we don't use this, so none of what you read is relevant."

Can we cut down on how much we talk about DTB. Since DTB is not in use, one would think we would be able to document our DT usage without even mentioning it.

Reviewer 3: FWIW, I like the DTB stuff. Gives some context when looking up device tree info elsewhere.

Author: > Can we cut down on how much we talk about DTB?

I mostly agree, though it could be interesting at least explaining why we're not using dtb the usual way and process dts at target compile time. This is not obvious to every one, I already had to explain this a couple of times to people used to Linux and that can't figure out the memory constraints we're facing.
So this is more targeting Linux users than non Linux aware developers, but still a useful piece of information. Anyway, we can try to shorten it a bit.

Currently supported boards
**************************

> Device tree is currently supported on all targets except posix (boards/posix).

Reviewer 2: xtensa?

Is this comment outdated?

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/dts.cmake#L26

Reviewer 2: Also, a reader might not be familiar with boarsd/posix.

So maybe instead say "Device tree is currently supported on all embedded targets".

Author: I found dts files in all extensa boards...

Author: > So maybe instead say "Device tree is currently supported on all embedded targets"

Well, it is supported by some qemu variants.
Sticking with posix and adding qemu_x86_64 whifch I initially forgot.

  dts/bindings/
  boards/<ARCH>/<BOARD>/

The common directory contains a skeleton.dtsi which provides device tree root
node definition.  The bindings subdirectory contains YAML binding files used

> to instruction how python DTS parsing script should extract nodes information

Reviewer 2: Something is wrong with the english:

"used to instruction how"

Author: Was supposed to be ""used to instruct how"". Txs

  dts/bindings/pinctrl/nxp,kinetis-pinmux.yaml
  dts/bindings/serial/nxp,kinetis-uart.yaml

YAML bindings for device nodes
******************************

Device tree describes hardware and configuration, but it doesn't tell the
system which pieces of information are useful, or how to generate configuration
data from the device tree nodes.  For this, we rely on YAML binding files to
describe the contents or definition of a device tree node and instruct how the
extracted information should be formatted.

A YAML description (called "YAML binding") must be provided for every device node
that is to be a source of information for the system.  A YAML binding file

> is associated to each node `compatible` property.  Information within the YAML

Author: @dbkinder , 'compatible' is a property of a node, such as 'label' is another possiebl property.
I don't understand what a node-compatible thing would be.

  dts/bindings/pinctrl/nxp,kinetis-pinmux.yaml
  dts/bindings/serial/nxp,kinetis-uart.yaml

YAML bindings for device nodes
******************************

Device tree describes hardware and configuration, but it doesn't tell the
system which pieces of information are useful, or how to generate configuration
data from the device tree nodes.  For this, we rely on YAML binding files to
describe the contents or definition of a device tree node and instruct how the
extracted information should be formatted.

> A YAML description (a.k.a YAML binding) must be provided for every device node

Reviewer 2: Do you need to introduce the term "YAML description" ?

Can we stick to "YAML binding" ? Or is there a difference?

Author: Actually YAML description is what it is, YAML bindings is how we call them.
Indeed, "a.k.a" makes it unclear. Let me reword it.

that is to be a source of information for the system.  A YAML binding file
is associated to each node `compatible` property.  Information within the YAML
file will instruct the python dts parsing script (located in `scripts/dts`) how
each property of the node is expected to be generated, either the type of the

> value or the format of it's name.  For now, node properties are generated as

Reviewer 2: Drop "For now," all documentation is "For now".

And being able to access node properties through defines is not likely to go away any time soon.

c-style #define in the format of usual include files that are included by the
system and made available to all zephyr components.

In a next step, YAML bindings could  also be used to validate the device tree
files themselves.  A device tree file can successfully compile and still not
contain the necessary pieces required to build the rest of the software.  YAML

> binding files provide a means to detect that issue in the future.

Reviewer 2: Drop this paragraph.

We should not be documenting what we would like the system to do. This is more appropriate in issues etc.

The user is reading this to understand what it does do.

compatible property.

Include files generation
************************

> At build, after board .dts file has been processed by dtc (device tree compiler),

Reviewer 4: ```suggestion
At build time, after a board's ``.dts`` file has been processed by the DTC (Device Tree Compiler),
```

> a frdm_k64f.dts_compiled file is generated under the zephyr directory.

Reviewer 4: ```suggestion
a corresponding ``.dts_compiled`` file is generated under the ``zephyr`` directory.
```

> This file is parsed by python dts parsing script, and thanks to YAML bindings it

Reviewer 4: How about:
```
This ``.dts_compiled`` file is processed by the python DTS parsing script
and generates an include file named
``include/generated/generated_dts_board_unfixed.h``
that holds all the information extracted from the DTS file with
the format specified by the YAML bindings.  For example:
```

generates, in the build output include/generated directory, an include file
named generated_dts_board_unfixed.h.
This include file holds all the information extracted from the dts file under
the format specified by the YAML bindings:

.. code-block:: c

   /* gpio_keys */
   #define DT_GPIO_KEYS_0		1

   /* button_0 */
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER	"GPIO_2"
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS	0
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_PIN		6
   #define DT_GPIO_KEYS_BUTTON_0_LABEL		"User SW2"
   #define DT_GPIO_KEYS_SW1_GPIO_CONTROLLER	DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER
   #define DT_GPIO_KEYS_SW1_GPIO_FLAGS		DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS
   #define DT_GPIO_KEYS_SW1_GPIO_PIN		DT_GPIO_KEYS_BUTTON_0_GPIO_PIN
   #define DT_GPIO_KEYS_SW1_LABEL			DT_GPIO_KEYS_BUTTON_0_LABEL
   #define SW1_GPIO_CONTROLLER			DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER
   #define SW1_GPIO_FLAGS				DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS
   #define SW1_GPIO_PIN				DT_GPIO_KEYS_BUTTON_0_GPIO_PIN
   #define SW1_LABEL				DT_GPIO_KEYS_BUTTON_0_LABEL

> Additionally, a concatenated file named generated_dts_board_fixups.h is

Reviewer 4: ```suggestion
Additionally, a file named ``generated_dts_board_fixups.h`` is
```

> generated in the same directory using all board related fixup files.

Reviewer 4: ```suggestion
generated in the same directory concatenating all board-related fixup files.
```

> The include file include/generated_dts_board.h includes both these generated

Reviewer 4: ```suggestion
The include file ``include/generated_dts_board.h`` includes both these generated
```

> files and hence gives access to DT information to Zephyr C source files.

Reviewer 4: ```suggestion
files, giving Zephyr C source files access to the board's device tree information.
```

generates, in the include/generated directory, an include file named
generated_dts_board_unfixed.h.
This include file holds all the information extracted from the dts file under
the format specified by the YAML bindings:

.. code-block:: c

   /* gpio_keys */
   #define DT_GPIO_KEYS_0		1

   /* button_0 */
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER	"GPIO_2"
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS	0
   #define DT_GPIO_KEYS_BUTTON_0_GPIO_PIN		6
   #define DT_GPIO_KEYS_BUTTON_0_LABEL		"User SW2"
   #define DT_GPIO_KEYS_SW1_GPIO_CONTROLLER	DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER
   #define DT_GPIO_KEYS_SW1_GPIO_FLAGS		DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS
   #define DT_GPIO_KEYS_SW1_GPIO_PIN		DT_GPIO_KEYS_BUTTON_0_GPIO_PIN
   #define DT_GPIO_KEYS_SW1_LABEL			DT_GPIO_KEYS_BUTTON_0_LABEL
   #define SW1_GPIO_CONTROLLER			DT_GPIO_KEYS_BUTTON_0_GPIO_CONTROLLER
   #define SW1_GPIO_FLAGS				DT_GPIO_KEYS_BUTTON_0_GPIO_FLAGS
   #define SW1_GPIO_PIN				DT_GPIO_KEYS_BUTTON_0_GPIO_PIN
   #define SW1_LABEL				DT_GPIO_KEYS_BUTTON_0_LABEL

Additionally, a concatenated file named generated_dts_board_fixups.h is
generated in the same directory using all board related fixup files.

> These both files are then combined into a generated_dts_board.conf file.

Reviewer 2: > These two files are included by both files are then combined into a generated_dts_board.conf file.

No it's not.

.conf corresponds to generated_dts_board_unfixed.h.

You could however say that:

"The source file generated_dts_board.h should be included by C source files to access DT information as it in turn includes generated_dts_board_unfixed.h and generated_dts_board_fixups.h.

Author: My bad. Txs for proposal.
